### PR TITLE
[6.2.0] Construct TreeArtifactValues on multiple threads.

### LINKED ISSUE #17009: Tree artifact up-to-dateness check can be very slow for large tree artifacts
### Description of the bug:

When checking whether a local action cache entry is up-to-date, it takes a long time to check actions that have large tree artifacts on their inputs. The stack trace when Bazel is working on this is:

```
   java.lang.Thread.State: RUNNABLE
        at java.io.FileInputStream.readBytes(java.base@11.0.6/Native Method)
        at java.io.FileInputStream.read(java.base@11.0.6/Unknown Source)
        at com.google.common.io.ByteStreams.copy(ByteStreams.java:114)
        at com.google.common.io.ByteSource.copyTo(ByteSource.java:257)
        at com.google.common.io.ByteSource.hash(ByteSource.java:340)
        at com.google.devtools.build.lib.vfs.FileSystem.getDigest(FileSystem.java:339)
        at com.google.devtools.build.lib.unix.UnixFileSystem.getDigest(UnixFileSystem.java:452)
        at com.google.devtools.build.lib.vfs.Path.getDigest(Path.java:690)
        at com.google.devtools.build.lib.vfs.DigestUtils.manuallyComputeDigest(DigestUtils.java:194)
        at com.google.devtools.build.lib.skyframe.ActionMetadataHandler.constructFileArtifactValue(ActionMetada
taHandler.java:564)
        at com.google.devtools.build.lib.skyframe.ActionMetadataHandler.constructFileArtifactValueFromFilesyste
m(ActionMetadataHandler.java:496)
        at com.google.devtools.build.lib.skyframe.ActionMetadataHandler.lambda$constructTreeArtifactValueFromFi
lesystem$0(ActionMetadataHandler.java:354)
        at com.google.devtools.build.lib.skyframe.ActionMetadataHandler$$Lambda$1121/0x0000000800857040.visit(Unknown Source)
        at com.google.devtools.build.lib.skyframe.TreeArtifactValue.visitTree(TreeArtifactValue.java:411)
        at com.google.devtools.build.lib.skyframe.TreeArtifactValue.visitTree(TreeArtifactValue.java:414)
        at com.google.devtools.build.lib.skyframe.TreeArtifactValue.visitTree(TreeArtifactValue.java:414)
        at com.google.devtools.build.lib.skyframe.TreeArtifactValue.visitTree(TreeArtifactValue.java:414)
        at com.google.devtools.build.lib.skyframe.TreeArtifactValue.visitTree(TreeArtifactValue.java:414)
        at com.google.devtools.build.lib.skyframe.TreeArtifactValue.visitTree(TreeArtifactValue.java:414)
        at com.google.devtools.build.lib.skyframe.TreeArtifactValue.visitTree(TreeArtifactValue.java:414)
        at com.google.devtools.build.lib.skyframe.TreeArtifactValue.visitTree(TreeArtifactValue.java:393)
        at com.google.devtools.build.lib.skyframe.ActionMetadataHandler.constructTreeArtifactValueFromFilesystem(ActionMetadataHandler.java:342)
        at com.google.devtools.build.lib.skyframe.ActionMetadataHandler.getTreeArtifactValue(ActionMetadataHandler.java:317)
        at com.google.devtools.build.lib.skyframe.ActionMetadataHandler.getMetadata(ActionMetadataHandler.java:265)
        at com.google.devtools.build.lib.actions.ActionCacheChecker.getMetadataOrConstant(ActionCacheChecker.java:566)
        at com.google.devtools.build.lib.actions.ActionCacheChecker.getMetadataMaybe(ActionCacheChecker.java:579)
        at com.google.devtools.build.lib.actions.ActionCacheChecker.validateArtifacts(ActionCacheChecker.java:207)
        at com.google.devtools.build.lib.actions.ActionCacheChecker.mustExecute(ActionCacheChecker.java:541)
```

My theory is that this is because the visitation happens on a single thread in `TreeArtifactValue.visitTree()` when called from `ActionMetadataHandler.constructTreeArtifactValueFromFilesystem()`.

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Take this BUILD file:

```
touch WORKSPACE
mkdir -p r
cat > r/BUILD <<'EOF'
load(":r.bzl", "r")

r(name = "ta")

genrule(
    name = "c",
    srcs = [":ta"],
    outs = ["co"],
    cmd = "find $(location :ta) > $@",
)

sh_binary(
    name = "gen",
    srcs = ["gen.sh"],
)
EOF

cat > r/r.bzl << 'EOF'
def _r_impl(ctx):
    ta = ctx.actions.declare_directory("d")
    ctx.actions.run(
        outputs = [ta],
        inputs = [],
        executable = ctx.executable._gen,
        arguments = [ta.path],
    )
    return [DefaultInfo(files = depset([ta]))]

r = rule(
    implementation = _r_impl,
    attrs = {
        "_gen": attr.label(default = "//r:gen", executable = True, cfg = "exec"),
    },
)
EOF

cat > r/gen.sh <<'EOF'
#!/bin/bash

OUT="$1"
mkdir -p "$OUT"

for i in $(seq 1 10); do
  for j in $(seq 1 10); do
    for k in $(seq 1 100); do
      mkdir -p "$OUT/$i/$j"
      #echo "$i $j $k" > "$OUT/$i/$j/$k"
      dd if=/dev/random of="$OUT/$i/$j/$k" bs=1024 count=1024
    done
  done
done
echo hello > "$OUT/hello"
EOF

chmod +x r/gen.sh

bazel build //r:c
bazel shutdown
bazel build //r:c  # This is slow

```

### Which operating system are you running Bazel on?

Linux @ Google

### What is the output of `bazel info release`?

development version

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

From git commit de4746da2394e49d028b5327ee5e65c6b60011cc .

### What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?

_No response_

### Have you found anything relevant by searching the web?

_No response_

### Any other information, logs, or outputs that you want to share?

_No response_

### PR DESCRIPTION
**Pull request by BalestraPatrick**
This makes it possible to use every core available for checksumming, which makes a huge difference for large tree artifacts.

Fixes #17009.

RELNOTES: None.
PiperOrigin-RevId: 525085502
Change-Id: I2a995d3445940333c21eeb89b4ba60887f99e51b (cherry picked from commit 368bf117513731909337b6e2bd5082c046399506)

## Context 
This is an attempt at cherry-picking this change to 6.2.0 as described in this comment: https://github.com/bazelbuild/bazel/commit/368bf117513731909337b6e2bd5082c046399506#commitcomment-110335567

Not totally sure if this is the right approach or there are more parts missing, but at least it built locally for me.

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/skyframe/TreeArtifactValue.java [hunk 1]
```diff
@@ -502,33 +571,18 @@ public interface TreeArtifactVisitor {
...
-
-      if (type == Dirent.Type.DIRECTORY) {
-        visitTree(parentDir, parentRelativePath, visitor);
-      }
+  public static void visitTree(Path parentDir, TreeArtifactVisitor treeArtifactVisitor) throws IOException {
+    try {
+      Visitor visitor = new Visitor(parentDir, treeArtifactVisitor);
+      visitor.run();
+    } catch (InterruptedException e) {
+      throw new IOException(e);
```
Comment (BalestraPatrick) [inline src/main/java/com/google/devtools/build/lib/skyframe/TreeArtifactValue.java:585]: The key part that is different is really only this one compared to https://github.com/bazelbuild/bazel/commit/368bf117513731909337b6e2bd5082c046399506#commitcomment-110473970

Comment (lberki) [reply src/main/java/com/google/devtools/build/lib/skyframe/TreeArtifactValue.java:585]: Yep, looks reasonable. Let's merge this!

**Review by lberki — APPROVED** [review]

**Review by keertk — APPROVED** [review]

**Comment by chiragramani** [comment]
Branch: release-6.2.0rc2
Commit: https://github.com/bazelbuild/bazel/commit/96eb991242746dc62a7cf85a07da26db09e5bcf2

We are seeing a regression on 6.2.0rc2 and looks like it might be related to this change. 

**Bazel 6.1.0**
Screenshot 2023-05-05 at 4 41 51 PM


- actuallyCompleteAction: 1.3seconds

**Bazel 6.2.0**
Screenshot 2023-05-05 at 4 42 51 PM

- actuallyCompleteAction: 6.7seconds

**Comment by keertk** [comment]
@chiragramani would you mind opening an issue for this so we can track it separately?

**Comment by BalestraPatrick** [comment]
I did 3 tries before this change and after, and here are the results of how long the `actuallyCompleteAction` in our bundling step of our app took (which is a tree artifact).

Before: 9.3s, 6.4s, 9.8s
After: 4.2s, 5.5s, 4.5s

So in our case we definitely see an improvement after this change. All measurements were performed by applying the same incremental modification to a single module, causing it to link and bundle the app again  I’ve noticed a big variance in these measurements from our developers as well after being on this change for a few days. Sometimes for the same exact change it can take 20s to compute the metadata, and other times it can take 4s.

**Comment by meisterT** [comment]
I would hope that once we have migrated to JDK20, and start using loom things like this become a lot easier to manage. cc @coeuvre

**Pull request merged**